src/pmdas/rocestat/pmdarocestat.python

```diff
@@ -0,0 +1,646 @@
+        nic = cache.contents
+
+        if 0 <= item < len(NicStats._fields_):
+            return [getattr(nic, item_lookup[item]), 1]
+        else:
+            return [c_api.PM_ERR_INST, 0]
+
+    def refresh_mlnx_counters(self):
+        ''' Refresh the values and instances for mlnx devices'''
+        try:
+            roce_stat = MLNXCOUNTERS()
+            hw_cnts = roce_stat.fetch_mlnx_counters()
+        except:
+            self.log("failed to fetch mlnx counters")
+
+        cached_mlnx_counters = {}
+        if bool(self.mlnx_counters):
+            for dev_name, subdict in self.mlnx_counters.items():
+                if not dev_name.endswith('delta'):
+                    prev_values = {}
+                    for counter_tuple in RoceDev._fields_:
+                        prev_values[counter_tuple[0]] = getattr(subdict, counter_tuple[0])
+
+                    cached_mlnx_counters[dev_name] = prev_values
+        else:
+            self.log("No counters in the cache")
+
+        hw_cnts_diff = {k: v for k, v in hw_cnts.items() if not k.endswith('delta')}
+
+        # if there are mlnx counters cached then calculate the delta
+        if bool(cached_mlnx_counters):
+            for key, subdict in cached_mlnx_counters.items():
+                hw_cnts_diff[f"{key}_delta"] = {}
+
+                temp = {}
+                for sub_key, value in subdict.items():
+                    temp[sub_key] = hw_cnts[key][sub_key] - cached_mlnx_counters[key][sub_key]
+
+                hw_cnts_diff[f"{key}_delta"] = temp
+        # If this is the first query then there wouldn't be cached counters
+        else:
+            for dev, subdict in hw_cnts.items():
+                temp = {k:0 for k in subdict}
+                hw_cnts_diff[dev+"_delta"] = temp
+
+        hw_cnts_diff = {k: dict(v.items()) for k, v in sorted(hw_cnts_diff.items())}
+
+        for device, dict_counters in hw_cnts_diff.items():
+            dev = RoceDev()
+            for k,v in dict_counters.items():
+                setattr(dev, k, v)
+
+            self.mlnx_counters[device] = dev
+
+    def refresh_nic_stats(self):
+        try:
+            collector = NICSTATS()
+            nic_stats = collector.collect_stats()
+        except:
+            self.log("failed to fetch nic stats")
+
+        cached_nic_stats = {}
+        if bool(self.nic_stats):
+            for dev_lane, subdict in self.nic_stats.items():
+                if not dev_lane.endswith('delta'):
+                    prev_values = {}
+                    for stat_tuple in NicStats._fields_:
+                        prev_values[stat_tuple[0]] = getattr(subdict, stat_tuple[0])
+
+                    cached_nic_stats[dev_lane] = prev_values
+        else:
+            self.log("No per lane stats in the cache")
+
+        bw = lambda x,y: float("%0.2f" % ((float(y)-float(x))*8.0e-9/deltat))
+        per_sec = lambda x,y: (int(x)-int(y))/deltat if deltat > 0 else (int(x)-int(y))
```

Comment:
Hmm, what is this all about?  This looks like a rate conversion - if so, this will be done in the PCP client tools automatically, and shouldn't usually be needed in the PMDA (as long as metrics are correctly exported as counters, which is not happening - see other review comments).